pkg/liqo-controller-manager/storageprovisioner/localprovisioner.go

```diff
@@ -0,0 +1,150 @@
+		return p.mutateLocalRealPVC(virtualPvc, &realPvc, options.SelectedNode)
+	}); err != nil {
+		return nil, controller.ProvisioningInBackground, err
+	} else if operation != controllerutil.OperationResultNone {
+		return nil, controller.ProvisioningInBackground, fmt.Errorf("provisioning real PVC")
+	}
+
+	if realPvc.Spec.VolumeName == "" {
+		return nil, controller.ProvisioningInBackground, fmt.Errorf("real PV not provided yet")
+	}
+
+	var realPv v1.PersistentVolume
+	if err := p.client.Get(ctx, types.NamespacedName{
+		Name: realPvc.Spec.VolumeName,
+	}, &realPv); err != nil {
+		return nil, controller.ProvisioningInBackground, err
+	}
+
+	pv := &v1.PersistentVolume{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: options.PVName,
+		},
+		Spec: v1.PersistentVolumeSpec{
+			StorageClassName: options.StorageClass.Name,
+			AccessModes:      options.PVC.Spec.AccessModes,
+			Capacity: v1.ResourceList{
+				v1.ResourceStorage: options.PVC.Spec.Resources.Requests[v1.ResourceStorage],
+			},
+			PersistentVolumeSource: realPv.Spec.PersistentVolumeSource,
+			NodeAffinity: &v1.VolumeNodeAffinity{
+				Required: &v1.NodeSelector{
+					NodeSelectorTerms: []v1.NodeSelectorTerm{
+						{
+							MatchExpressions: []v1.NodeSelectorRequirement{
+								{
+									Key:      consts.TypeLabel,
+									Operator: v1.NodeSelectorOpDoesNotExist,
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	if options.StorageClass.ReclaimPolicy != nil {
+		pv.Spec.PersistentVolumeReclaimPolicy = *options.StorageClass.ReclaimPolicy
+	}
+
+	pv.Spec.NodeAffinity = mergeAffinities(&pv.Spec, &realPv.Spec)
+
+	return pv, controller.ProvisioningFinished, nil
+}
+
+// mutateLocalRealPVC enforces the local realPVC fields merging its old values (if any)
+// with the ones coming from the virtualPVC.
+// i.e. the PVC spec is the copy of the virtual one, but the storage class is the one set in the
+// storage provisioner or the one previously set in the PVC (since it is a read-only field). The
+// real volumeName is preserved too.
+func (p *liqoLocalStorageProvisioner) mutateLocalRealPVC(virtualPvc, realPvc *v1.PersistentVolumeClaim,
+	selectedNode *v1.Node) error {
+	if realPvc.ObjectMeta.Annotations == nil {
+		realPvc.ObjectMeta.Annotations = map[string]string{}
+	}
+	// required if the real storage class is wait for first consumer
+	realPvc.ObjectMeta.Annotations["volume.kubernetes.io/selected-node"] = selectedNode.Name
+	realPvc.ObjectMeta.Annotations["volume.alpha.kubernetes.io/selected-node"] = selectedNode.Name
+
+	storageClassName := realPvc.Spec.StorageClassName
+	if p.localRealStorageClass != "" {
+		storageClassName = &p.localRealStorageClass
+	}
+
+	realPvName := realPvc.Spec.VolumeName
+	realPvc.Spec = *virtualPvc.Spec.DeepCopy()
+	realPvc.Spec.VolumeName = realPvName
+	realPvc.Spec.StorageClassName = storageClassName
+
+	return nil
+}
+
+func mergeAffinities(vol1, vol2 *v1.PersistentVolumeSpec) *v1.VolumeNodeAffinity {
```

Comment:
1. Would double-check whether a function similar to this one does not already exist (IIRC, the webhook is doing something very similar)
2. If not, I would add some unit tests to ensure its correctness, given the overall complexity.